source/codegen/templates/service_helpers.mako

```diff
@@ -41,22 +41,30 @@ ${initialize_input_params(function_name, parameters)}
 %>\
 ${initialize_input_params(function_name, non_ivi_params)}\
 
-      auto status = library_->${function_name}(${service_helpers.create_args_for_ivi_dance(parameters)});
-      if (status < 0) {
-        response->set_status(status);
-        return ::grpc::Status::OK;
-      }
-      ${size_param['type']} ${common_helpers.camel_to_snake(size_param['cppName'])} = status;
-
+      while (true) {
+        auto status = library_->${function_name}(${service_helpers.create_args_for_ivi_dance(parameters)});
+        if (status < 0) {
+          response->set_status(status);
+          return ::grpc::Status::OK;
+        }
+        ${size_param['type']} ${common_helpers.camel_to_snake(size_param['cppName'])} = status;
+      
+<%block filter="common_helpers.indent(1)">\
 ${initialize_output_params(output_parameters)}\
-      status = library_->${function_name}(${service_helpers.create_args(parameters)});
-      response->set_status(status);
+</%block>\
+        status = library_->${function_name}(${service_helpers.create_args(parameters)});
+        if (status == kErrorReadBufferTooSmall || status > ${common_helpers.camel_to_snake(size_param['cppName'])}) {
```

Comment:
From looking at some examples I _think_ that ivi-dance functions can either return the "buffer too small" error or just return the required buffer size. In either case to make the logic simple we go around the loop again.